blob/service.go

```diff
@@ -49,26 +49,111 @@ type Submitter interface {
 }
 
 type Service struct {
+	// ctx represents the Service's lifecycle context.
+	ctx    context.Context
+	cancel context.CancelFunc
 	// accessor dials the given celestia-core endpoint to submit blobs.
 	blobSubmitter Submitter
 	// shareGetter retrieves the EDS to fetch all shares from the requested header.
 	shareGetter share.Getter
 	// headerGetter fetches header by the provided height
 	headerGetter func(context.Context, uint64) (*header.ExtendedHeader, error)
+	// headerSub subscribes to new headers to supply to blob subscriptions.
+	headerSub func(ctx context.Context) (<-chan *header.ExtendedHeader, error)
 }
 
 func NewService(
 	submitter Submitter,
 	getter share.Getter,
 	headerGetter func(context.Context, uint64) (*header.ExtendedHeader, error),
+	headerSub func(ctx context.Context) (<-chan *header.ExtendedHeader, error),
 ) *Service {
 	return &Service{
 		blobSubmitter: submitter,
 		shareGetter:   getter,
 		headerGetter:  headerGetter,
+		headerSub:     headerSub,
 	}
 }
 
+func (s *Service) Start(context.Context) error {
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+	return nil
+}
+
+func (s *Service) Stop(context.Context) error {
+	s.cancel()
+	return nil
+}
+
+type SubscriptionResponse struct {
+	Blobs  []*Blob
+	Height uint64
+}
+
+func (s *Service) Subscribe(ctx context.Context, ns share.Namespace) (<-chan *SubscriptionResponse, error) {
+	if s.ctx == nil {
+		return nil, fmt.Errorf("service has not been started")
+	}
+
+	headerCh, err := s.headerSub(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	blobCh := make(chan *SubscriptionResponse, 16)
+	go func() {
+		defer close(blobCh)
+
+		for {
+			select {
+			case header, ok := <-headerCh:
+				if ctx.Err() != nil {
+					log.Debug("blobsub: canceling subscription due to user context closing")
+					return
+				}
+				if !ok {
+					log.Errorw("header channel closed for subscription", "namespace", ns.ID())
+					return
+				}
+				// close subscription before buffer overflows
+				if len(blobCh) == cap(blobCh)-1 {
+					return
+				}
+
+				var blobs []*Blob
+				var err error
+				for {
+					blobs, err = s.getAll(ctx, header, []share.Namespace{ns})
+					if err == nil || ctx.Err() != nil {
+						// operation successful, break the loop
+						break
+					}
+				}
+				if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+					// context canceled, continuing would lead to unexpected missed heights for the client
+					log.Debug("blobsub: canceling subscription due to user context closing")
+					return
+				}
```

Comment:
* Do context check once and return immediately
* Check for any ctx error, rather than two specific, as there might be more added.

```suggestion
				var blobs []*Blob
				var err error
				for {
					blobs, err = s.getAll(ctx, header, []share.Namespace{ns})
					if ctx.Err() != nil {
						// context canceled, continuing would lead to unexpected missed heights for the client
						log.Debug("blobsub: canceling subscription due to user context closing")
						return
					}
					if err == nil ||  {
						// operation successful, break the loop
						break
					}
				}
	
```